tslearn/nonmyopic.py

```diff
@@ -0,0 +1,509 @@
+
+    def clustering(self, X):
+        """
+        This function compute and fit clusters for a training dataset of time series using K-means and euclidean
+        distances
+        Parameters
+        ----------
+        X: array-like
+            The training dataset of time series
+
+        Returns
+        -------
+        Vector : a vector as long as the number of series in the training dataset that indicates in which cluster the
+        series at the same index belongs
+
+         Float : the silhouette score of the fitting of the clustering
+        """
+        self.cluster_ = TimeSeriesKMeans(n_clusters=self.number_cluster)
+        c_k = self.cluster_.fit_predict((to_time_series_dataset(X)))
+        silhouette = silhouette_score(
+            to_time_series_dataset(X), c_k, metric="euclidean"
+        )
+
+        return c_k, silhouette
+
+    def fit(self, X, Y):
+        """
+        This function fits classifiers that are currently multilayer perceptrons to a training set of time series and
+        associated classes. A classifier is fit for each time stamp above a minimum time stamp that is an attribute of
+        the class of the model. Then some probabilities are computed using clusters already computed.
+        This function should be divided among 2 or 3 functions.
+
+        Parameters
+        ----------
+        X: Array-like
+            a dataset of time series
+        Y: vector
+            the associated classes of the series from X_train
+        """
+        classes_y = np.unique(Y)
+        c_k, silhouette = self.clustering(X)
+        self.silhouette_ = silhouette
+        self.classifier_ = []
+        self.clusters_ = {}
+        self.pyhatyck_ = {}
+        self.indice_ck_ = []
+        self.number_classes_ = len(classes_y)
+        self.len_X_ = X.shape[1]
+        mid_X = int(X.shape[0] / 2)
+        X1 = X[:mid_X]
+        Y1 = Y[:mid_X]
+        X2 = X[mid_X:]
+        Y2 = Y[mid_X:]
+        c_k2 = c_k[mid_X:]
+        self.pyck_ = np.zeros(shape=(self.number_classes_, self.number_cluster))
+        for current_cluster in range(0, self.number_cluster):
+            self.clusters_["ck_cm{0}".format(current_cluster)] = []
+            self.pyhatyck_["pyhatycks{0}".format(current_cluster)] = []
+            self.indice_ck_.append(np.where(c_k == current_cluster))
+            Y_current_cluster = Y[self.indice_ck_[-1]]
+            for current_class in range(0, self.number_classes_):
+                value_current_class = classes_y[current_class]
+                if len(Y_current_cluster) == 0:
+                    current_pyck = 0
+                else:
+                    current_pyck = len(
+                        Y_current_cluster[Y_current_cluster == value_current_class]
+                    ) / len(Y_current_cluster)
+                self.pyck_[current_class][current_cluster] = current_pyck
+
+        for i in range(self.minimum_time_stamp, X.shape[1] + 1):
```

Comment:
Please use variable names that reflect their use, eg. :
* `i` is for time series index
* `t` is for time index
* `di` is for dimension index
* `k` is for cluster index